protobuf/decode.go

```diff
@@ -0,0 +1,282 @@
+package protobuf
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/Fantom-foundation/Substate/substate"
+	"github.com/Fantom-foundation/Substate/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+type dbGetCode = func(types.Hash) ([]byte, error)
+
+// Decode converts protobuf-encoded Substate into aida-comprehensible substate
+func (s *Substate) Decode(lookup dbGetCode, block uint64, tx int) (*substate.Substate, error) {
+	input, err := s.GetInputAlloc().decode(lookup)
+	if err != nil {
+		return nil, err
+	}
+
+	output, err := s.GetOutputAlloc().decode(lookup)
+	if err != nil {
+		return nil, err
+	}
+
+	environment, err := s.GetBlockEnv().decode()
+	if err != nil {
+		return nil, err
+	}
+
+	message, err := s.GetTxMessage().decode(lookup)
+	if err != nil {
+		return nil, err
+	}
+
+	contractAddress := s.GetTxMessage().getContractAddress()
+	result, err := s.GetResult().decode(contractAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	return &substate.Substate{
+		InputSubstate:  *input,
+		OutputSubstate: *output,
+		Env:            environment,
+		Message:        message,
+		Result:         result,
+		Block:          block,
+		Transaction:    tx,
+	}, nil
+}
+
+// decode converts protobuf-encoded Substate_Alloc into aida-comprehensible WorldState
+func (alloc *Substate_Alloc) decode(lookup dbGetCode) (*substate.WorldState, error) {
+	world := make(substate.WorldState, len(alloc.GetAlloc()))
+
+	for _, entry := range alloc.GetAlloc() {
+		addr, acct, err := entry.decode()
+		if err != nil {
+			return nil, fmt.Errorf("Error decoding alloc entry; %w", err)
+		}
+
+		address := types.BytesToAddress(addr)
+		nonce, balance, _, codehash, err := acct.decode()
+		if err != nil {
+			return nil, fmt.Errorf("Error decoding entry account; %w", err)
+		}
+
+		code, err := lookup(codehash)
+		if err != nil && !errors.Is(err, leveldb.ErrNotFound) {
+			return nil, fmt.Errorf("Error looking up codehash; %w", err)
+		}
```

Comment:
I guess this is from old version and we do want to continue even if `code` is not found, right?

Reply:
From old version - Yes.

This is the only way I have found that would validate record/replay pair correctly. 
That is, when `codehash` is not part of the lookup, validation expects a `nil`.